test/integration/api/authentication.spec.js

```diff
@@ -49,8 +112,122 @@ const authorizationServerInfo = {
+        }
+      }
+    })
+    .expect(200)
+
+  const browser = await puppeteer.launch({
+    executablePath: '/usr/bin/chromium-browser',
+    args: ['--disable-dev-shm-usage', '--no-sandbox']
+  })
+
+  const page = await browser.newPage()
+
+  // Chrome on Linux
+  await page.setUserAgent('Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.103 Safari/537.36')
+
+  await executeBrowserScenario({ browser, page })
+
+  await new Promise(resolve => setTimeout(resolve, 1000))
+
+  const afterLoginRedirectUrl = page.url()
+  t.true(afterLoginRedirectUrl.startsWith(afterAuthenticationUrl))
+
+  const applicationCodeUrlObj = new URL(afterLoginRedirectUrl)
+  t.is(typeof applicationCodeUrlObj.searchParams.get('code'), 'string')
+  t.is(applicationCodeUrlObj.searchParams.get('status'), 'success')
+
+  const { body: obj } = await request(t.context.serverUrl)
+    .post('/auth/token')
+    .set({
+      'x-platform-id': t.context.platformId,
+      'x-stelace-env': t.context.env
+    })
+    .send({
+      code: applicationCodeUrlObj.searchParams.get('code'),
+      grantType: 'authorizationCode'
+    })
+    .expect(200)
+
+  t.is(typeof obj, 'object')
+  t.is(obj.tokenType, 'Bearer')
+  t.is(typeof obj.accessToken, 'string')
+  t.is(typeof obj.refreshToken, 'string')
+  t.truthy(obj.userId)
+
+  const token = jwt.decode(obj.accessToken)
+  t.is(typeof token, 'object')
+  t.is(typeof token.loggedAt, 'number')
+  t.deepEqual(token.roles, ['user', 'provider'])
+  t.truthy(token.userId)
+
+  // Tokens can only be fetched once
+  const { body: getTokenError } = await request(t.context.serverUrl)
+    .post('/auth/token')
+    .set({
+      'x-platform-id': t.context.platformId,
+      'x-stelace-env': t.context.env
+    })
+    .send({
+      code: applicationCodeUrlObj.searchParams.get('code'),
+      grantType: 'authorizationCode'
+    })
+    .expect(422)
+
+  t.true(getTokenError.message.includes('already used'))
+
+  const ssoAuthorizationHeaders = {
+    'x-platform-id': t.context.platformId,
+    'x-stelace-env': t.context.env,
+    authorization: `${obj.tokenType} ${obj.accessToken}`
+  }
+
+  // fetch the created user
+  const { body: user } = await request(t.context.serverUrl)
+    .get(`/users/${token.userId}`)
+    .set(ssoAuthorizationHeaders)
+    .expect(200)
+
+  t.deepEqual(user.roles, ['user', 'provider'])
+  t.true(user.platformData.ssoProviders.includes(provider))
+
+  await browser.close()
+}
+
 test.before(async t => {
-  await before({ name: 'authentication' })(t)
+  // set fixed values to be able to test OAuth authentication with external providers
+  // with stable URLs
+  // App URL: http://localhost:9461
+  // Authorization callback URL: http://localhost:7645/auth/sso/e8_test/${provider}/callback
+  // Current tested providers: 'github', 'facebook'
+  process.env.SERVER_PORT = 7645
+  const oauthServerPort = 9461
+
+  await before({
+    name: 'authentication',
+    platformId: 8,
```

Comment:
Why platform 8. Can you mention platforms currently used as well (like platform 9?).

Reply:
Currently besides platform 9, the platform 1 is reserved for instant seed and 2 for api key tests.

https://github.com/stelace/stelace/blob/654ea5afc38bc4f99595edce01bde6d4e433d51c/scripts/instantData.js#L15
https://github.com/stelace/stelace/blob/654ea5afc38bc4f99595edce01bde6d4e433d51c/test/integration/api/apiKey.spec.js#L15

Is it better if I set it as platform 3 ?

Reply:
Yes please, it’s easier to keep track of used test platforms this way.